src/Giraffe/WebSockets.fs

```diff
@@ -0,0 +1,182 @@
+                    let cancellationToken = cancellationToken |> Option.defaultValue CancellationToken.None
+                    do! this.WebSocket.SendAsync(segment, WebSocketMessageType.Text, true, cancellationToken)
+        }
+        
+
+        /// Closes the connection to the WebSocket client.
+        member this.CloseAsync(?reason,?cancellationToken) = task {
+            if not (isNull this.WebSocket) then
+                if this.WebSocket.State = WebSocketState.Open then
+                    let cancellationToken = cancellationToken |> Option.defaultValue CancellationToken.None
+                    let reason = reason |> Option.defaultValue "Closed by the WebSocket server"
+                    do! this.WebSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, reason, cancellationToken)
+        }
+
+        
+        /// Creates a new reference to a WebSocket.
+        static member FromWebSocket(websocket,?webSocketID,?subProtocol) : WebSocketReference = {
+            WebSocket  = websocket
+            Subprotocol = subProtocol
+            ID = webSocketID |> Option.defaultWith (fun _ -> Guid.NewGuid().ToString())
+        }
+
+/// A connection manager keeps track of all connections that are open at a specific endpoint.
+type ConnectionManager(?messageSize) =
+    let messageSize = defaultArg messageSize DefaultWebSocketOptions.ReceiveBufferSize
+
+    let connections = new System.Collections.Concurrent.ConcurrentDictionary<string, WebSocketReference>()
+
+    with
+
+        /// Tries to find a WebSocket by ID. Returns None if the socket wasn't found.
+        member __.TryGetWebSocket(websocketID:string) : WebSocketReference option = 
+            match connections.TryGetValue websocketID with
+            | true, r -> Some r
+            | _ -> None
+
+        /// Returns the number of WebSocket connections.
+        member __.Count = connections.Count
+
+        /// Sends a UTF-8 encoded text message to all WebSocket connections.
+        member __.BroadcastTextAsync(msg:string,?cancellationToken:CancellationToken) = task {
+            let byteResponse = System.Text.Encoding.UTF8.GetBytes msg
+            let segment = ArraySegment<byte>(byteResponse, 0, byteResponse.Length)
+            let cancellationToken = cancellationToken |> Option.defaultValue CancellationToken.None
+
+            let! _ =
+                connections
+                |> Seq.map (fun kv -> task {
+                    try
+                        if not cancellationToken.IsCancellationRequested then
+                            let webSocket = kv.Value.WebSocket
+                            if webSocket.State = WebSocketState.Open then
+                                do! webSocket.SendAsync(segment, WebSocketMessageType.Text, true, cancellationToken)
+                            else
+                                if webSocket.State = WebSocketState.Closed then
+                                    connections.TryRemove kv.Key |> ignore
```

Comment:
Removing element from a collection that's being enumerated... I don't know if it's OK for ConcurrentDictionary, but it's a strong code smell.

Reply:
I agree, this would definitely fail on the normal dictionary but due to locks in Concurrent version, might be preventing usual error. Might be cleaner/safer to add a temp "IdsToRemove" list that is then enumerated after the initial run to remove.

Reply:
done